File: Anurag_19_CSE/Python/day24.py
# Link For Problem: https://leetcode.com/problems/longest-valid-parentheses/

class Solution:

    """ 
        TC : O(n)

        SC : O(n)
    """

    def longestValidParentheses(self, s: str) -> int:
        ans: int = 0
        stack: list[int] = []

        stack.append(-1)

        for i, c in enumerate(s):

            if c == '(':
                stack.append(i)

            else:
                stack.pop()

                if not stack:
                    stack.append(i)

                else:
                    ans = max(ans, i-stack[-1])

        return ans

File: Anurag_19_CSE/Java/day24.java
//Link For Problem: https://leetcode.com/problems/longest-valid-parentheses/

import java.util.ArrayDeque;

public class day24 {

    /*
     * TC : O(n)
     * 
     * SC : O(n)
     */

    public int longestValidParentheses(String s) {
        int ans = 0;
        ArrayDeque<Integer> dq = new ArrayDeque<>();

        dq.offerFirst(-1);

        for (int i = 0; i < s.length(); i++) {

            if (s.charAt(i) == '(')
                dq.offerFirst(i);

            else {
                dq.pollFirst();

                if (dq.isEmpty())
                    dq.offerFirst(i);

                else
                    ans = Math.max(ans, i - dq.peekFirst());
            }
        }

        return ans;
    }
}
